cider/common/hashtable/FixedHashMap.h

```diff
@@ -0,0 +1,189 @@
+
+  const VoidKey getKey() const { return {}; }  /// NOLINT
+  Mapped& getMapped() { return mapped; }
+  const Mapped& getMapped() const { return mapped; }
+
+  bool isZero(const State&) const { return !mapped; }
+  void setZero() { mapped = {}; }
+
+  /// Similar to FixedHashSetCell except that we need to contain a pointer to the Mapped
+  /// field.
+  ///  Note that we have to assemble a continuous layout for the value_type on each call
+  ///  of getValue().
+  struct CellExt {
+    CellExt() {}  //-V730 /// NOLINT
+    CellExt(Key&& key_, const FixedHashMapImplicitZeroCell* ptr_)
+        : key(key_), ptr(const_cast<FixedHashMapImplicitZeroCell*>(ptr_)) {}
+    void update(Key&& key_, const FixedHashMapImplicitZeroCell* ptr_) {
+      key = key_;
+      ptr = const_cast<FixedHashMapImplicitZeroCell*>(ptr_);
+    }
+    Key key;
+    FixedHashMapImplicitZeroCell* ptr;
+
+    const Key& getKey() const { return key; }
+    Mapped& getMapped() { return ptr->mapped; }
+    const Mapped& getMapped() const { return ptr->mapped; }
+    const value_type getValue() const { return {key, ptr->mapped}; }  /// NOLINT
+  };
+};
+
+template <typename Key,
+          typename Mapped,
+          typename Cell = FixedHashMapCell<Key, Mapped>,
+          typename Size = FixedHashTableStoredSize<Cell>,
+          typename Allocator = HashTableAllocator>
+class FixedHashMap : public FixedHashTable<Key, Cell, Size, Allocator> {
+ public:
+  using Base = FixedHashTable<Key, Cell, Size, Allocator>;
+  using Self = FixedHashMap;
+  using LookupResult = typename Base::LookupResult;
+
+  using Base::Base;
+
+  template <typename Func, bool>
+  void ALWAYS_INLINE mergeToViaEmplace(Self& that, Func&& func) {
+    for (auto it = this->begin(), end = this->end(); it != end; ++it) {
+      typename Self::LookupResult res_it;
+      bool inserted;
+      that.emplace(it->getKey(), res_it, inserted, it.getHash());
+      func(res_it->getMapped(), it->getMapped(), inserted);
+    }
+  }
+
+  template <typename Func>
+  void ALWAYS_INLINE mergeToViaFind(Self& that, Func&& func) {
+    for (auto it = this->begin(), end = this->end(); it != end; ++it) {
+      auto res_it = that.find(it->getKey(), it.getHash());
+      if (!res_it)
+        func(it->getMapped(), it->getMapped(), false);
+      else
+        func(res_it->getMapped(), it->getMapped(), true);
+    }
+  }
+
+  template <typename Func>
+  void forEachValue(Func&& func) {
+    for (auto& v : *this)
+      func(v.getKey(), v.getMapped());
+  }
+
+  template <typename Func>
+  void forEachMapped(Func&& func) {
+    for (auto& v : *this)
+      func(v.getMapped());
+  }
+
+  Mapped& ALWAYS_INLINE operator[](const Key& x) {
+    LookupResult it;
+    bool inserted;
+    this->emplace(x, it, inserted);
+    if (inserted)
+      new (&it->getMapped()) Mapped();
+
+    return it->getMapped();
+  }
+};
+
+template <typename Key, typename Mapped, typename Allocator = HashTableAllocator>
+using FixedImplicitZeroHashMap =

```

Comment:
`FixedImplicitZeroHashMap` `FixedImplicitZeroHashMapWithCalculatedSize` `FixedHashMap` are all interface you will provide? what's the difference and how to choose. Do we need ut for these basic class and API? Please provide more info via comments.

Reply:
Yes, `FixedImplicitZeroHashMap` is for int8 and `FixedImplicitZeroHashMapWithCalculatedSize` is for int16. Let me add some comments here.